File: android/src/main/java/com/capacitor/mediaviewer/MediaViewerPlugin.java
package com.capacitor.mediaviewer;

import android.app.Activity;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.List;
import java.util.ArrayList;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "MediaViewer")
public class MediaViewerPlugin extends Plugin {

    private MediaViewerFragment mediaViewerFragment;
    private MediaViewerListener mediaViewerListener;

    @Override
    public void load() {
        mediaViewerListener = new MediaViewerListener() {
            @Override
            public void onPlaybackStateChanged(PlaybackState state) {
                notifyListeners("playbackStateChanged", state.toJSObject());
            }

            @Override
            public void onMediaIndexChanged(int index) {
                JSObject data = new JSObject();
                data.put("index", index);
                notifyListeners("mediaIndexChanged", data);
            }

            @Override
            public void onViewerDismissed() {
                notifyListeners("viewerDismissed", new JSObject());
            }
        };
    }

    @PluginMethod
    public void show(PluginCall call) {
        try {
            JSONObject options = call.getData();
            JSONArray itemsArray = options.getJSONArray("items");
            int currentIndex = options.getInt("currentIndex");
            String title = options.optString("title", "");

            List<MediaItem> mediaItems = parseMediaItems(itemsArray);

            Activity activity = getActivity();
            if (activity == null) {
                call.reject("Activity is null");
                return;
            }

            if (!(activity instanceof FragmentActivity)) {
                call.reject("Activity must be a FragmentActivity");
                return;
            }

            FragmentActivity fragmentActivity = (FragmentActivity) activity;
            fragmentActivity.runOnUiThread(() -> {
                mediaViewerFragment = MediaViewerFragment.newInstance(
                    mediaItems,
                    currentIndex,
                    title,
                    mediaViewerListener
                );
                FragmentManager fragmentManager = fragmentActivity.getSupportFragmentManager();
                mediaViewerFragment.show(
                    fragmentManager,
                    "MediaViewerFragment"
                );
                call.resolve();
            });
        } catch (JSONException e) {
            call.reject("Error parsing options: " + e.getMessage());
        }
    }

    @PluginMethod
    public void dismiss(PluginCall call) {
        Activity activity = getActivity();
        if (activity != null && mediaViewerFragment != null) {
            activity.runOnUiThread(() -> {
                mediaViewerFragment.dismiss();
                mediaViewerFragment = null;
                call.resolve();
            });
        } else {
            call.resolve();
        }
    }

    @PluginMethod
    public void play(PluginCall call) {
        if (mediaViewerFragment != null) {
            mediaViewerFragment.play();
            call.resolve();
        } else {
            call.reject("Media viewer is not showing");
        }
    }

    @PluginMethod
    public void pause(PluginCall call) {
        if (mediaViewerFragment != null) {
            mediaViewerFragment.pause();
            call.resolve();
        } else {
            call.reject("Media viewer is not showing");
        }
    }

    @PluginMethod
    public void seek(PluginCall call) {
        try {
            double time = call.getDouble("time");
            if (mediaViewerFragment != null) {
                mediaViewerFragment.seek((long) (time * 1000)); // Convert to milliseconds
                call.resolve();
            } else {
                call.reject("Media viewer is not showing");
            }
        } catch (Exception e) {
            call.reject("Error seeking: " + e.getMessage());
        }
    }

    @PluginMethod
    public void setQuality(PluginCall call) {
        try {
            String quality = call.getString("quality");
            if (mediaViewerFragment != null) {
                mediaViewerFragment.setQuality(quality);
                call.resolve();
            } else {
                call.reject("Media viewer is not showing");
            }
        } catch (Exception e) {
            call.reject("Error setting quality: " + e.getMessage());
        }
    }

    @PluginMethod
    public void getPlaybackState(PluginCall call) {
        if (mediaViewerFragment != null) {
            PlaybackState state = mediaViewerFragment.getPlaybackState();
            call.resolve(state.toJSObject());
        } else {
            call.reject("Media viewer is not showing");
        }
    }

    private List<MediaItem> parseMediaItems(JSONArray itemsArray) throws JSONException {
        List<MediaItem> items = new ArrayList<>();
        for (int i = 0; i < itemsArray.length(); i++) {
            JSONObject itemObj = itemsArray.getJSONObject(i);
            MediaItem item = new MediaItem();
            item.url = itemObj.getString("url");
            item.type = itemObj.getString("type");
            item.title = itemObj.optString("title", null);

            if (itemObj.has("qualityVariants")) {
                JSONArray qualityArray = itemObj.getJSONArray("qualityVariants");
                item.qualityVariants = new ArrayList<>();
                for (int j = 0; j < qualityArray.length(); j++) {
                    JSONObject qualityObj = qualityArray.getJSONObject(j);
                    QualityVariant variant = new QualityVariant();
                    variant.label = qualityObj.getString("label");
                    variant.url = qualityObj.getString("url");
                    item.qualityVariants.add(variant);
                }
            } else if ("video".equals(item.type) && HlsPlaylistParser.isHlsUrl(item.url)) {
                // Automatically parse HLS master playlist for quality variants
                new Thread(() -> {
                    List<QualityVariant> variants = HlsPlaylistParser.parseMasterPlaylist(item.url);
                    if (variants != null && !variants.isEmpty()) {
                        item.qualityVariants = variants;
                    }
                }).start();
            }

            items.add(item);
        }
        return items;
    }
}

File: ios/Plugin/MediaViewerPlugin.swift
import Foundation
import Capacitor
import AVKit
import AVFoundation
import UIKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(MediaViewerPlugin)
public class MediaViewerPlugin: CAPPlugin {
    private var mediaViewerController: MediaViewerController?
    
    @objc func show(_ call: CAPPluginCall) {
        guard let itemsArray = call.getArray("items", [String: Any].self) else {
            call.reject("Items array is required")
            return
        }
        
        guard let currentIndex = call.getInt("currentIndex") else {
            call.reject("Current index is required")
            return
        }
        
        let title = call.getString("title") ?? ""
        
        var mediaItems: [MediaItem] = []
        for itemDict in itemsArray {
            guard let pathString = itemDict["path"] as? String,
                  let typeString = itemDict["type"] as? String else {
                continue
            }
            
            let item = MediaItem()
            item.path = pathString
            item.type = typeString
            item.alt = itemDict["alt"] as? String
            item.thumbnail = itemDict["thumbnail"] as? String
            
            mediaItems.append(item)
        }
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.mediaViewerController = MediaViewerController(
                mediaItems: mediaItems,
                currentIndex: currentIndex,
                title: title,
                plugin: self
            )
            
            if let viewController = self.bridge?.viewController {
                self.mediaViewerController?.modalPresentationStyle = .fullScreen
                viewController.present(self.mediaViewerController!, animated: true) {
                    call.resolve()
                }
            } else {
                call.reject("View controller not available")
            }
        }
    }
    
    @objc func dismiss(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.mediaViewerController?.dismiss(animated: true) {
                self.mediaViewerController = nil
                call.resolve()
            }
        }
    }
    
    @objc func play(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.mediaViewerController?.play()
            call.resolve()
        }
    }
    
    @objc func pause(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.mediaViewerController?.pause()
            call.resolve()
        }
    }
    
    @objc func seek(_ call: CAPPluginCall) {
        guard let time = call.getDouble("time") else {
            call.reject("Time is required")
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            self?.mediaViewerController?.seek(to: time)
            call.resolve()
        }
    }
    
    @objc func setQuality(_ call: CAPPluginCall) {
        guard let quality = call.getString("quality") else {
            call.reject("Quality is required")
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            self?.mediaViewerController?.setQuality(quality)
            call.resolve()
        }
    }
    
    @objc func getPlaybackState(_ call: CAPPluginCall) {
        guard let state = mediaViewerController?.getPlaybackState() else {
            call.reject("Media viewer is not showing")
            return
        }
        
        var result: [String: Any] = [
            "isPlaying": state.isPlaying,
            "currentTime": state.currentTime,
            "duration": state.duration
        ]
        
        if let quality = state.currentQuality {
            result["currentQuality"] = quality
        }
        
        call.resolve(result)
    }
    
    func notifyPlaybackStateChanged(_ state: PlaybackState) {
        var data: [String: Any] = [
            "isPlaying": state.isPlaying,
            "currentTime": state.currentTime,
            "duration": state.duration
        ]
        
        if let quality = state.currentQuality {
            data["currentQuality"] = quality
        }
        
        notifyListeners("playbackStateChanged", data: data)
    }
    
    func notifyMediaIndexChanged(_ index: Int) {
        notifyListeners("mediaIndexChanged", data: ["index": index])
    }
    
    func notifyViewerDismissed() {
        notifyListeners("viewerDismissed", data: [:])
    }
}

